cpp/examples/arrow/exec_plan_examples.cc

```diff
@@ -0,0 +1,1122 @@
+                                ::arrow::internal::GetCpuThreadPool());
+
+    // ensure arrow::dataset node factories are in the registry
+    arrow::dataset::internal::Initialize();
+
+    // Execution plan created
+    ARROW_ASSIGN_OR_RAISE(
+        std::shared_ptr<cp::ExecPlan> plan, cp::ExecPlan::Make(&exec_context));
+
+    std::shared_ptr<arrow::dataset::Dataset> dataset = CreateDataset();
+
+    auto options = std::make_shared<arrow::dataset::ScanOptions>();
+    // sync scanning is not supported by ScanNode
+    options->use_async = true;
+    options->projection = Materialize({});  // create empty projection
+
+    // construct the scan node
+    cp::ExecNode* scan;
+    auto scan_node_options = arrow::dataset::ScanNodeOptions{dataset, options};
+
+    ARROW_ASSIGN_OR_RAISE(scan,
+                            cp::MakeExecNode("scan", plan.get(), {}, scan_node_options));
+
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+    cp::ExecNode* sink;
+
+    ARROW_ASSIGN_OR_RAISE(
+        sink, cp::MakeExecNode("sink", plan.get(), {scan},
+        cp::SinkNodeOptions{&sink_gen}));
+
+    // // translate sink_gen (async) to sink_reader (sync)
+    std::shared_ptr<arrow::RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+        dataset->schema(), std::move(sink_gen), exec_context.memory_pool());
+
+    // validate the ExecPlan
+    ABORT_ON_FAILURE(plan->Validate());
+    PRINT_LINE("ExecPlan created : " << plan->ToString());
+    // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+
+    // // collect sink_reader into a Table
+    std::shared_ptr<arrow::Table> response_table;
+
+    ARROW_ASSIGN_OR_RAISE(response_table,
+                            arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+    PRINT_LINE("Results : " << response_table->ToString());
+
+    // // stop producing
+    plan->StopProducing();
+    // // plan mark finished
+    plan->finished().Wait();
+
+    return arrow::Status::OK();
+}
+
+cp::ExecBatch GetExecBatchFromJSON(const std::vector<arrow::ValueDescr>& descrs,
+                                   arrow::util::string_view json) {
+  auto fields = ::arrow::internal::MapVector(
+      [](const arrow::ValueDescr& descr) { return arrow::field("", descr.type); },
+      descrs);
+
+  cp::ExecBatch batch{*GetRecordBatchFromJSON(arrow::schema(std::move(fields)), json)};
+
+  auto value_it = batch.values.begin();
+  for (const auto& descr : descrs) {
+    if (descr.shape == arrow::ValueDescr::SCALAR) {
+      if (batch.length == 0) {
+        *value_it = arrow::MakeNullScalar(value_it->type());
+      } else {
+        *value_it = value_it->make_array()->GetScalar(0).ValueOrDie();
+      }
+    }
+    ++value_it;
+  }
+
+  return batch;
+}
+
+struct BatchesWithSchema {
+  std::vector<cp::ExecBatch> batches;
+  std::shared_ptr<arrow::Schema> schema;
+
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> gen(bool parallel) const {
```

Comment:
Would it make more sense to demonstrate InMemoryDataset here? This is using a lot of semi-internal machinery that I think will be confusing.

Reply:
One question related to this, the `SourceNodeOptions` requires the input

```c++
SourceNodeOptions(std::shared_ptr<Schema> output_schema,
                    std::function<Future<util::optional<ExecBatch>>()> generator) {...}
```

Are you referring to the return type of this function or about the internal components used in the function body.

Reply:
Ah, right. A `Dataset` can be turned into a `Scanner` and the `Scanner` can _internally_ build an ExecPlan, but that isn't exposed. However, that means we're stuck using internal APIs in an example which we should avoid as well.

Reply:
What is the best option for this? Not quite sure how to wrap this function?

Reply:
Sorry, I had thought there was a public conversion from Dataset/Scanner to ExecNode…it seems not. I pinged Weston below. However here I would want to make at least two changes:

1. Make it clear that this implementation is for ease of demonstration, not performance
2. Get rid of the `parallel` parameter and only demonstrate the non-parallel path (since it's an example anyways)

Reply:
Removed `parallel` option.

Reply:
> Can we add a placeholder section and link to that perhaps?
> […](#)
> On Tue, Jan 4, 2022, at 11:48, Vibhatha Lakmal Abeykoon wrote: ***@***.**** commented on this pull request. In docs/source/cpp/streaming_execution.rst <[#12033 (comment)](https://github.com/apache/arrow/pull/12033#discussion_r778232832)>: > + // data schema + auto schema = arrow::schema({...}) + + // source node options + auto source_node_options = arrow::compute::SourceNodeOptions{schema, gen}; + + // create a source node + ARROW_ASSIGN_OR_RAISE(arrow::compute::ExecNode * source, + arrow::compute::MakeExecNode("source", plan.get(), {}, + source_node_options)); + +``FilterNode`` +-------------- + +:class:`FilterNode`, as the name suggests, provide a container to define a data filtering criteria. +Filter can be written using :class:`arrow::compute::Expression`. For instance if the row values compute.rst contains a some of the function definitions, but not a detailed examples on how to use expressions. — Reply to this email directly, view it on GitHub <[#12033 (comment)](https://github.com/apache/arrow/pull/12033#discussion_r778232832)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AACQB37JP4IFUAYPITMT3JLUUMQFPANCNFSM5KUSEHEA>. Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. You are receiving this because you were mentioned.Message ID: ***@***.***>

@lidavidm should we create a new RST or just a sub-section within the same document?

Reply:
Hmm, it probably goes in its own section of compute.rst. That said, it occurs to me now that unless we intentionally fill out the section before 7.0.0, it would be rather confusing, and linking to the API docs for now (or just not linking if there's nothing to link to…?) is better.

Reply:
I can work on a separate PR for it and may be once it is reviewed properly we can link to this. 
I guess it won't take too much time. @westonpace @lidavidm WDYT?

Reply:
I think I mentioned this below but the "scan" node is the correct way IMO to go from dataset + scan options -> exec node.  However, "ScanOptions" is something of a hidden class (especially in python) and there is a JIRA to hide it completely, exposing only ScannerBuilder so that is a bit of a conundrum.  Maybe ScannerBuilder could have a BuildExecNode method which creates a scan node.  @lidavidm thoughts?

Reply:
For now I think it is fine to document everything but scan and we can add scan later in a follow-up PR once we align on this.